src/api/index.ml

```diff
@@ -116,21 +116,119 @@ let read_val file =
+    the values (element of ocaml language). First element is value's name and second
     is its type. *)
 
+let read_types file =
+  begin
+  match EzFile.read_lines_to_list file with
+  | _ ::
+    _ ::
+    _ ::
+    mdl_types ->
+    (* Translation [identifier_x1;type_nature_y1;signature_z1;x2;y2;z2...] to [(x1,y1,z1);(x2,y2,z2);...] *)
+      let rec separate acc = function
+        | [] -> List.rev acc
+        | ident::typ::sign ->
+
+            let get_signature_and_continue =
+              let rec aux_get_sig acc =
+                function
+                  | [] -> List.rev acc, []
+                  | x::r ->
+                      match x with
+                      | "" -> List.rev acc, r
+                      | s -> aux_get_sig (s::acc) r
+                in fun file_sig -> aux_get_sig [] file_sig
+              in
+
+            let signature, rest_of_file =
+              get_signature_and_continue sign
+            in
+            separate ((ident, typ, signature)::acc) rest_of_file
+
+        | _s ->
+            Format.printf "SEPARATING TYPES FOR %s" file;
+            failwith "should not occur"
+      in
+        separate [] mdl_types
+  | _lines ->
+    Printf.eprintf "Unrecognized format for TYPE_MODULE file %S\n%!" file;
+      raise Not_found
+  end
+(** Reads meta file ('TYPES.MODULE._') and returns list of triplets that describe
+    the type signatures (element of ocaml language).
+    First element is type's name,
+    Second is the nature of its type (TYPE_ABSTRACT, TYPE_CLASS, TYPE_VARIANT,...).
+      This might be brought to change at some point for no
+    third is the actual signature. *)
+
+let type_id_cnt =
+  let count = ref (-1) in
+  fun () ->
+    incr count;
+    !count
+
+let list_type_declarations mdl file =
+
+  let init_type mdl ident =
+      let type_id = Int32.of_int (type_id_cnt ()) in
+      {type_id; ident; constructors = ["..."];}
+  in
+
+  let classes, types =
+    List.partition (fun (ident, typkind, typesig) ->
+      match typkind with
+      | "CLASS_OBJ" -> true
+
+      | "TYPE_ABSTRACT" | "TYPE_OPEN" | "TYPE_RECORD" | "TYPE_VARIANT" -> false
+
+      | _s -> Format.eprintf "TYPEKIND ERROR in file %s@.Type declaration preceding: %s \"%s\" (%a)" file typkind ident
+      ( Format.pp_print_list ~pp_sep:(fun fmt () -> Format.fprintf fmt " ") Format.pp_print_string ) typesig;
+failwith "TYPES.m type kind should not occur"
+      ) (read_types file)
+  in
+
+  let types =
+    List.map (fun (ident, typkind, typesig) ->
+      let given_type = init_type mdl ident in
+      match typkind with
+      | "TYPE_OPEN" -> given_type
+      | "TYPE_ABSTRACT" | "TYPE_RECORD" | "TYPE_VARIANT" ->
+          { given_type with constructors = typesig }
+      | _s -> assert false
+    ) types in
+    mdl.mdl_types <- types;
+
+    let classes =
+      List.map (fun (ident, typkind, typesig) ->
+        let given_type = init_type mdl ident in
```

Comment:
I'm afraid separating the count between types and classes might bring duplicate keys in the table though

Reply:
since you store classes and types in different tables, and its not referenced anywhere in the same column, it won't happen. But, it's not necessary, so as you like:)